src/http/params.cr

```diff
@@ -328,6 +328,21 @@ module HTTP
       end
     end
 
+    # Returns a copy of this `HTTP::Params` instance
+    #
+    # ```
+    # require "http/params"
+    #
+    # original = HTTP::Params{"name" => "Jamie"}
+    # updated = params.dup
+    # updated["name"] = "Ary"
+    #
+    # original["name"] # => "Jamie"
+    # ```
+    def dup : self
+      self.class.new(raw_params.dup)
```

Comment:
But raw_params is `Hash(String, Array(String))` a `dup` on it should leave aliasing to the underlying Array.

The spec works because `Params#[]=(name, value : String | Array(String))` in master is creating a new Array on every invocation when the value is a String. This was changed in #9605.

The following behavior seems wrong with this PR

```crystal
original = HTTP::Params{"name" => ["Jamie"]}
updated = original.dup
updated.add("name", "Ary")
updated.fetch_all("name")  # => ["Jamie", "Ary"]
original.fetch_all("name") # => ["Jamie", "Ary"]
```

I think dup should clone directly in this case.

Reply:
Ahh, good point. I’ll add a test case for this.

> I think dup should clone directly in this case.

Oh, is `clone` the convention for deep-duping an object?

Reply:
dup means get another instance with the same underlying values. If you dup an array that holds arrays, those arrays are exactly the same ones. clone does a deep clone.

So it might be fine to keep the same arrays when duping, at least that's the semantic. You would use clone to get different inner arrays. But for HTTP::Params I don't know what's a use case for dup.

Reply:
> dup means get another instance with the same underlying values. If you dup an array that holds arrays, those arrays are exactly the same ones.

This is a good point. I suppose I expected it to work like a hash:

```crystal
require "http"

params = HTTP::Params { "foo" => "bar" }
# => HTTP::Params(@raw_params={"foo" => ["bar"]}) : HTTP::Params
hash = { "foo" => "bar" }
# => {"foo" => "bar"} : Hash(String, String)

params.dup["foo"] = "lol"
hash.dup["foo"] = "lol"

params # => HTTP::Params(@raw_params={"foo" => ["lol"]}) : HTTP::Params
hash   # => {"foo" => "bar"} : Hash(String, String)
```

The fact that `HTTP::Params` wraps a hash is an implementation detail that I wouldn't expect to be exposed. I also can't see how it'd even be useful to dup the `Params` without duping the hash. 🙂 So I wouldn't expect `dup` to keep the same hash, but I suppose keeping the arrays actually makes sense with that context because the hash would do the same thing, right?

> But for HTTP::Params I don't know what's a use case for dup.

I'm generating links based on params for the current request. Something like this (but not necessarily exactly this):

```erb
<a href="?<%= r.query_params.dup.merge!({ "page" => page - 1 }) %>">Previous</a>
<a href="?<%= r.query_params.dup.merge!({ "page" => page + 1 }) %>">Next</a>
```

Reply:
I think your implementation is fine. The default `dup` for structs just returns the same struct because structs are passed by value. However, maybe `dup` for structs should create a new instance where all instance vars are duped by default.

With your implementation `params` is `HTTP::Params(@raw_params={"foo" => ["bar"]})` in your snippet above.

Reply:
>  However, maybe `dup` for structs should create a new instance where all instance vars are duped by default.

That's an interesting thought. I wonder what the performance difference would be in real-world apps. It's reasonable to assume it would have _some_ performance impact, but I wonder if it would be noticeable in most apps.

On the flip side, I wonder if there are cases where duping ivars could cause issues, like if a struct referenced an `IO` or a connection pool (maybe a DB-backed model), I imagine duplicating those references could get weird.

Reply:
@bcardiff Given this discussion, I think it actually makes sense _not_ to dup the array. Do you have any thoughts on that?

Reply:
What I still hold it does not make sense it to have two `HTTP::Params` instances that share the underlying hash or array.

In that aspect the `def initialize(@raw_params : Hash(String, Array(String)))` is unsafe. I would rather stay on the safe side here. HTTP::Params are usually not big structures.

If there are some memory concerns, in the future the representation could be change to `@raw_params : Hash(String, String|Array(String))` for example.

Regarding passing arguments, this type been a struct that is actually a wrapper to a reference type _that never changes_, it ends up acting as a reference type. So there effectively more than one http params instance pointing to the same Hash through assignments or arguments passing. That is fine to me, but sharing the underlying Array is bad. It can trigger weird bugs.